File: services/notification-service/src/notification/notification.service.ts
import { Injectable, Logger, NotFoundException, BadRequestException } from '@nestjs/common';
import { randomUUID } from 'node:crypto';
import { createTransport, type Transporter } from 'nodemailer';
import { WebClient } from '@slack/web-api';

import { ConfigService } from '@nestjs/config';
import { DatabaseService } from '../database/database.service';
import { RequestContextService } from '../observability/request-context.service';
import {
  CreateTemplateDto,
  ListNotificationQueryDto,
  NotificationChannel,
  SendNotificationDto,
  SubscribeDto,
  WebhookReceiptDto,
} from './dto/notification.dto';

export interface NotificationRow {
  id: string;
  channel: string;
  recipient: string;
  subject: string | null;
  body: string;
  status: string;
  created_at: Date;
  updated_at: Date;
}

export interface TemplateRow {
  id: string;
  name: string;
  subject: string | null;
  body: string;
  channel: string;
  created_at: Date;
}

export interface SubscriberRow {
  id: string;
  url: string;
  tenant_id: string | null;
  created_at: Date;
}

@Injectable()
export class NotificationService {
  private readonly logger = new Logger(NotificationService.name);
  private emailTransport: Transporter;
  private slackClient: WebClient;

  constructor(
    private readonly db: DatabaseService,
    private readonly config: ConfigService,
    private readonly requestContext: RequestContextService,
  ) {
    const smtp = this.config.get<{
      host: string;
      port: number;
      user: string;
      pass: string;
      from: string;
    }>('smtp')!;
    this.emailTransport = createTransport({
      host: smtp.host,
      port: smtp.port,
      secure: smtp.port === 465,
      auth: smtp.user ? { user: smtp.user, pass: smtp.pass } : undefined,
    });

    const slackToken = this.config.get<string>('slack.token') as string;
    this.slackClient = new WebClient(slackToken);
  }

  async send(dto: SendNotificationDto): Promise<NotificationRow> {
    let subject = dto.subject ?? '';
    let body = dto.body ?? '';

    if (dto.template) {
      const tpl = await this.findTemplate(dto.template, dto.channel);
      subject = tpl.subject ?? subject;
      body = this.interpolate(tpl.body, dto.data ?? {});
    }

    if (!body && !subject) {
      throw new BadRequestException('Notification requires body, subject, or a template');
    }

    const id = randomUUID();
    const result = await this.db.query<NotificationRow>(
      `INSERT INTO notifications (id, channel, recipient, subject, body, status)
       VALUES ($1, $2, $3, $4, $5, 'pending') RETURNING *`,
      [id, dto.channel, dto.recipient, subject, body],
    );
    const row = result.rows[0]!;

    try {
      await this.deliver(dto.channel, dto.recipient, subject, body);
      await this.markStatus(id, 'sent');
      row.status = 'sent';
    } catch (err) {
      this.logger.error({ err, id }, 'Notification delivery failed');
      await this.markStatus(id, 'failed');
      row.status = 'failed';
    }

    await this.notifySubscribers(row);
    return row;
  }

  private async deliver(channel: string, recipient: string, subject: string, body: string): Promise<void> {
    switch (channel) {
      case NotificationChannel.EMAIL:
        await this.emailTransport.sendMail({
          from: this.config.get<string>('smtp.from'),
          to: recipient,
          subject,
          text: body,
        });
        break;
      case NotificationChannel.SLACK: {
        const webhook = this.config.get<string>('slack.webhookUrl');
        if (webhook) {
          await fetch(webhook, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ text: body, channel: recipient }),
          });
        } else {
          await this.slackClient.chat.postMessage({ channel: recipient, text: body });
        }
        break;
      }
      case NotificationChannel.PUSH: {
        await this.db.query(
          `INSERT INTO notification_device_tokens (token, payload) VALUES ($1, $2::jsonb)
           ON CONFLICT (token) DO UPDATE SET payload = EXCLUDED.payload`,
          [recipient, JSON.stringify({ subject, body })],
        );
        break;
      }
      default:
        throw new BadRequestException(`Unsupported channel: ${channel}`);
    }
  }

  async list(query: ListNotificationQueryDto) {
    const page = query.page ?? 1;
    const pageSize = query.pageSize ?? 20;
    const offset = (page - 1) * pageSize;

    const conditions: string[] = [];
    const params: unknown[] = [];
    let idx = 1;
    if (query.channel) {
      conditions.push(`channel = $${idx}`);
      params.push(query.channel);
      idx++;
    }
    const where = conditions.length ? `WHERE ${conditions.join(' AND ')}` : '';
    const countRes = await this.db.query<{ count: string }>(
      `SELECT COUNT(*) as count FROM notifications ${where}`,
      params,
    );
    const total = Number(countRes.rows[0]!.count);
    const data = await this.db.query<NotificationRow>(
      `SELECT * FROM notifications ${where} ORDER BY created_at DESC LIMIT $${idx} OFFSET $${idx + 1}`,
      [...params, pageSize, offset],
    );
    return { data: data.rows, meta: { page, pageSize, total, totalPages: Math.ceil(total / pageSize) } };
  }

  async get(id: string): Promise<NotificationRow> {
    const result = await this.db.query<NotificationRow>(
      `SELECT * FROM notifications WHERE id = $1`,
      [id],
    );
    if (result.rows.length === 0) {
      throw new NotFoundException(`Notification "${id}" not found`);
    }
    return result.rows[0]!;
  }

  async createTemplate(dto: CreateTemplateDto): Promise<TemplateRow> {
    const id = randomUUID();
    const result = await this.db.query<TemplateRow>(
      `INSERT INTO notification_templates (id, name, subject, body, channel)
       VALUES ($1, $2, $3, $4, $5) RETURNING *`,
      [id, dto.name, dto.subject ?? null, dto.body, dto.channel],
    );
    return result.rows[0]!;
  }

  async listTemplates(): Promise<TemplateRow[]> {
    const result = await this.db.query<TemplateRow>(
      `SELECT * FROM notification_templates ORDER BY created_at DESC`,
    );
    return result.rows;
  }

  async subscribe(dto: SubscribeDto): Promise<SubscriberRow> {
    const id = randomUUID();
    const result = await this.db.query<SubscriberRow>(
      `INSERT INTO notification_subscribers (id, url, tenant_id)
       VALUES ($1, $2, $3) RETURNING *`,
      [id, dto.url, dto.tenantId ?? null],
    );
    return result.rows[0]!;
  }

  async receiveReceipt(id: string, dto: WebhookReceiptDto): Promise<void> {
    await this.markStatus(id, dto.status ?? 'unknown');
    this.logger.log({ id, status: dto.status }, 'Delivery receipt received');
  }

  private async findTemplate(name: string, channel: string): Promise<TemplateRow> {
    const result = await this.db.query<TemplateRow>(
      `SELECT * FROM notification_templates WHERE name = $1 AND channel = $2`,
      [name, channel],
    );
    if (result.rows.length === 0) {
      throw new NotFoundException(`Template "${name}" for channel "${channel}" not found`);
    }
    return result.rows[0]!;
  }

  private async markStatus(id: string, status: string): Promise<void> {
    await this.db.query(
      `UPDATE notifications SET status = $1, updated_at = NOW() WHERE id = $2`,
      [status, id],
    );
  }

  private async notifySubscribers(notification: NotificationRow): Promise<void> {
    const subs = await this.db.query<SubscriberRow>(
      `SELECT * FROM notification_subscribers`,
    );
    await Promise.all(
      subs.rows.map((sub) =>
        fetch(sub.url, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify(notification),
        }).catch((err) => this.logger.warn({ err, url: sub.url }, 'Subscriber webhook failed')),
      ),
    );
  }

  private interpolate(template: string, data: Record<string, unknown>): string {
    return template.replace(/\{\{\s*(\w+)\s*\}\}/g, (match, key: string) => {
      const value = data[key];
      return value !== undefined ? String(value) : match;
    });
  }
}

File: apps/backend/src/backend/services/notification_service.py
import uuid
import re
from datetime import datetime, timezone

import httpx
from fastapi import HTTPException
from sqlalchemy import select, text
from sqlalchemy.ext.asyncio import AsyncSession

from ..models.schema import Notification


class NotificationService:
    async def send(self, dto, db: AsyncSession = None):
        now = datetime.now(timezone.utc)
        body = dto.body
        subject = dto.subject

        if dto.template:
            tpl = await self.resolve_template(dto.template, dto.channel.value, db)
            if tpl:
                subject = subject or tpl.get("subject", "")
                body = body or tpl.get("body", "")
                if dto.data:
                    body = self.interpolate_template(body, dto.data)
                    if subject:
                        subject = self.interpolate_template(subject, dto.data)

        if not body:
            raise HTTPException(400, "No body resolved")

        notification = Notification(
            type="outgoing",
            channel=dto.channel.value,
            title=subject or "",
            message=body,
            recipient=dto.recipient,
            subject=subject,
            status="pending",
            priority="medium",
        )
        db.add(notification)
        await db.flush()

        try:
            if dto.channel.value == "email":
                notification.status = "sent"
            elif dto.channel.value == "slack":
                notification.status = "sent"
            elif dto.channel.value == "push":
                notification.status = "sent"
            notification.updated_at = datetime.now(timezone.utc)
        except Exception:
            notification.status = "failed"
            notification.updated_at = datetime.now(timezone.utc)

        await db.flush()
        await db.refresh(notification)

        await self.notify_subscribers(notification, db)
        return notification

    async def list_notifications(self, page: int, page_size: int, channel: str | None, db: AsyncSession = None):
        query = select(Notification)
        if channel:
            query = query.where(Notification.channel == channel)
        query = query.order_by(Notification.created_at.desc())
        total_result = await db.execute(select(Notification.id).where(query._where_criteria[0] if query._where_criteria else True))
        total = len(total_result.all())
        offset = (page - 1) * page_size
        query = query.offset(offset).limit(page_size)
        result = await db.execute(query)
        rows = result.scalars().all()
        return rows, total

    async def get_notification(self, notification_id: uuid.UUID | str, db: AsyncSession = None):
        if isinstance(notification_id, str):
            notification_id = uuid.UUID(notification_id)
        result = await db.execute(
            select(Notification).where(Notification.id == notification_id)
        )
        return result.scalar_one_or_none()

    async def create_template(self, dto, db: AsyncSession = None):
        template_id = uuid.uuid4()
        now = datetime.now(timezone.utc)
        await db.execute(
            text("""
                INSERT INTO notification_templates (id, name, subject, body, channel, created_at)
                VALUES (:id, :name, :subject, :body, :channel, :created_at)
            """),
            {
                "id": template_id,
                "name": dto.name,
                "subject": dto.subject,
                "body": dto.body,
                "channel": dto.channel.value,
                "created_at": now,
            },
        )
        await db.flush()
        return {"id": str(template_id), "name": dto.name, "subject": dto.subject, "body": dto.body, "channel": dto.channel.value, "created_at": now}

    async def list_templates(self, db: AsyncSession = None):
        result = await db.execute(
            text("SELECT * FROM notification_templates ORDER BY created_at DESC")
        )
        rows = result.fetchall()
        return [dict(r._mapping) for r in rows]

    async def resolve_template(self, name: str, channel: str, db: AsyncSession = None):
        result = await db.execute(
            text("SELECT * FROM notification_templates WHERE name = :name AND channel = :channel"),
            {"name": name, "channel": channel},
        )
        row = result.fetchone()
        return dict(row._mapping) if row else None

    def interpolate_template(self, template: str, data: dict) -> str:
        def replacer(match):
            key = match.group(1)
            return str(data.get(key, match.group(0)))
        return re.sub(r"\{\{(\w+)\}\}", replacer, template)

    async def subscribe(self, dto, db: AsyncSession = None):
        sub_id = uuid.uuid4()
        now = datetime.now(timezone.utc)
        await db.execute(
            text("""
                INSERT INTO notification_subscribers (id, url, tenant_id, created_at)
                VALUES (:id, :url, :tenant_id, :created_at)
            """),
            {"id": sub_id, "url": dto.url, "tenant_id": dto.tenant_id, "created_at": now},
        )
        await db.flush()
        return {"id": str(sub_id), "url": dto.url, "tenant_id": dto.tenant_id, "created_at": now}

    async def notify_subscribers(self, notification, db: AsyncSession = None):
        try:
            result = await db.execute(text("SELECT url FROM notification_subscribers"))
            subscribers = result.fetchall()
            async with httpx.AsyncClient(timeout=5) as client:
                for sub in subscribers:
                    try:
                        await client.post(sub[0], json={
                            "id": str(notification.id),
                            "channel": notification.channel,
                            "recipient": notification.recipient,
                            "subject": notification.subject,
                            "body": notification.message,
                            "status": notification.status,
                        })
                    except Exception:
                        pass
        except Exception:
            pass

    async def update_status(self, notification_id: str, status: str, db: AsyncSession = None):
        result = await db.execute(
            text("UPDATE notifications SET status = :status, updated_at = :updated_at WHERE id = :id"),
            {"status": status, "updated_at": datetime.now(timezone.utc), "id": uuid.UUID(notification_id)},
        )
        await db.flush()
        return result.rowcount > 0


notification_service = NotificationService()
